Add missing methods to unix ExitStatusExt

ijackson opened the pull request:
These are the methods corresponding to the remaining exit status examination macros from `wait.h`.  `WCOREDUMP` isn't in SuS but is it is very standard.  I have not done portability testing to see if this builds everywhere, so I may need to Do Something if it doesn't.

There is also a bugfix and doc improvement to `.signal()`, and an `.into_raw()` accessor.

This would fix #73128 and fix #73129.  Please let me know if you like this direction, and if so I will open the tracking issue and so on.

If this MR goes well, I may tackle #73125 next - I have an idea for how to do it.

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

ijackson commented:
@rustbot modify labels +T-libs

See also the tracking issue #73131

jyn514 commented on library/std/src/sys/unix/ext/process.rs:
```diff
@@ -171,8 +171,32 @@ pub trait ExitStatusExt {
 
     /// If the process was terminated by a signal, returns that signal.
+    ///
+    /// Ie, if `WIFSIGNALED`, this returns `WTERMSIG`.
```
I think `i.e.` is not clear for non-native speakers (and maybe some native speakers).
```suggestion
    /// In other words, if `WIFSIGNALED`, this returns `WTERMSIG`.
```

jyn514 commented on library/std/src/sys/unix/ext/process.rs:
```diff
@@ -171,8 +171,32 @@ pub trait ExitStatusExt {
+
+    /// If the process was stopped by a signal, returns that signal.
+    ///
+    /// Ie, if `WIFSTOPPED`, this returns `WSTOPSIG`.  This is only possible if the status came from
```
```suggestion
    /// In other words, if `WIFSTOPPED`, this returns `WSTOPSIG`.  This is only possible if the status came from
```

jyn514 commented on library/std/src/sys/unix/process/process_unix.rs:
```diff
@@ -479,7 +479,23 @@ impl ExitStatus {
 
     pub fn signal(&self) -> Option<i32> {
-        if !self.exited() { Some(libc::WTERMSIG(self.0)) } else { None }
+        if libc::WIFSIGNALED(self.0) { Some(libc::WTERMSIG(self.0)) } else { None }
```
What's the difference between `exited()` and `IFSIGNALED`?

ijackson replied:
`!exited()` is `!WIFEXITED`, which is not the same as `WIFSIGNALED`.  That is, `WIFEXITED` and `WIFSIGNALED` are not exhaustive.  Some wait statuses are `!WIFSIGNALED && !WIFEXITED`.

The commit message for that commit explains this in a bit more detail, with the example of `WIFSTOPPED`

Mark-Simulacrum commented:
r? @dtolnay -- needs T-libs decision on the general direction. I am also not familiar enough with the details of signals and such to know whether this is correct (and my quick manpage investigation is not feeling too productive).

I did flag a concern that this is extending a non-sealed stable extension trait. I forget if we've previously decided that's acceptable, but I presume that would be a breaking change.

jyn514 commented:
> I did flag a concern that this is extending a non-sealed stable extension trait. I forget if we've previously decided that's acceptable, but I presume that would be a breaking change.

This broke async-std in https://github.com/rust-lang/rust/issues/77089, so I would expect this is a no go for that reason alone :/

ijackson commented:
> This broke async-std in #77089, so I would expect this is a no go for that reason alone :/

Urgh.  It hadn't occurred to me that anyone would implement this for their own types.  ISTM that not making this trait sealed was a serious mistake.

Given where we are now, I think our options are:
1. Add a sealed trait `std::os::unix::process::ExitStatusExtExt` (and maybe get rid of it again in the 2021 edition)
2. Retrospectively make this trait sealed, presumably with a crater run to check for problems.
3. Never extend the set of unix-specific methods available on `ExitStatus`

I think 3 is clearly unacceptable given the fact that, currently, this type is missing at the very least the absolutely necessary access to `WCOREDUMP` and the obviously desirable `into_raw`.  Without either `WCOREDUMP` or `into_raw` it is actually impossible to run a subcommand from Rust, and report its outcome toi the user fully correctly (ie, including whether it coredumped, like the shell does), without effedtively reimplementing the whole of `Command`.

`ExitStatusExtExt` is very ugly, but ugliness is the natural consequence of trying to fix earlier API design errors in a backward compatible way :-(.

Sealing the trait may or may not be disruptive.  If a crater run would suffice, it could be run on this branch as-is, since adding methods will break all other implementors.

In defence of the idea of retrospectively sealing `ExitStatusExt`, the situation with `ExitStatus` is rather different to `OpenOptions`.  `OpenOptions` has the `open` method - substantial functionality, which `async-std` needed to replace while retaining the rest of the methods.  `ExitStatus` doesn't contain any other functionality so I am hoping there wouldn't be the same kind of reason for implementing it for one's own types.

jyn514 commented:
> Sealing the trait may or may not be disruptive. If a crater run would suffice, it could be run on this branch as-is, since adding methods will break all other implementors.

A crater run seems like a good idea to me, but we should probably get feedback from T-libs that this is functionality they want to add first.

dtolnay commented on library/std/src/sys/unix/ext/process.rs:
```diff
@@ -171,8 +171,32 @@ pub trait ExitStatusExt {
+    ///
+    /// In other words, if `WIFSTOPPED`, this returns `WSTOPSIG`.  This is only possible if the status came from
+    /// a `wait` system call which was passed `WUNTRACED`, was then converted into an `ExitStatus`.
+    #[stable(feature = "rust1", since = "1.0.0")]
```
`#[stable(feature = "rust1", since = "1.0.0")]` is not correct since this wasn't in 1.0.0. I'd like to land these all as unstable first (like you have for `core_dumped` and `into_raw`).

ijackson replied:
This was a c&p mistake on my part.  Will fix, thanks.

dtolnay reviewed (changes requested):
I am on board with these. Please file a tracking issue and we can figure out how to land them.

I don't think a crater run is necessarily worth it for this. If someone has a quick way to grep all crates for the string `ExitStatusExt for` we can do that but otherwise I'm comfortable enough with https://github.com/search?l=Rust&o=desc&q=%22ExitStatusExt+for%22&s=indexed&type=Code showing no results outside std. I think we can land this and listen for reported breakage, and leave it to the first beta crater run to measure fallout.

However I'd like for both ExitStatusExt traits to be changed to sealed in the same PR.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1645152264) [(plain)](https://github.com/rust-lang/rust/commit/1a11796aacc49d8d8b1814056fd9b6b89c24e7c6/checks/1645152264/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking miniz_oxide v0.4.0
    Checking object v0.22.0
    Checking hashbrown v0.9.0
    Checking addr2line v0.14.0
error: expected one of `!`, `(`, `+`, `::`, `<`, `where`, or `{`, found `;`
   --> library/std/src/sys/unix/ext/process.rs:212:39
    |
212 | impl private::Sealed for ExitStatusExt;
    |                                       ^ expected one of 7 possible tokens

error[E0432]: unresolved imports `super::process::CommandExt`, `super::process::ExitStatusExt`
   |
   |
74 |     pub use super::process::{CommandExt, ExitStatusExt};
   |                              ^^^^^^^^^^  ^^^^^^^^^^^^^ no `ExitStatusExt` in `sys::unix::ext::process`
   |                              |
   |                              no `CommandExt` in `sys::unix::ext::process`

error[E0599]: no method named `as_raw_fd` found for reference `&ChildStdin` in the current scope
   --> library/std/src/sys/unix/kernel_copy.rs:364:44
    |
364 |         CopyParams(FdMeta::Pipe, Some(self.as_raw_fd()))
    |                                            ^^^^^^^^^ method not found in `&ChildStdin`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `sys::unix::ext::io::AsRawFd` defines an item `as_raw_fd`, perhaps you need to implement it
    |
    |
22  | pub trait AsRawFd {


error[E0599]: no method named `as_raw_fd` found for reference `&ChildStdout` in the current scope
   --> library/std/src/sys/unix/kernel_copy.rs:370:44
    |
370 |         CopyParams(FdMeta::Pipe, Some(self.as_raw_fd()))
    |                                            ^^^^^^^^^ method not found in `&ChildStdout`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `sys::unix::ext::io::AsRawFd` defines an item `as_raw_fd`, perhaps you need to implement it
    |
    |
22  | pub trait AsRawFd {


error[E0599]: no method named `as_raw_fd` found for reference `&ChildStderr` in the current scope
   --> library/std/src/sys/unix/kernel_copy.rs:376:44
    |
376 |         CopyParams(FdMeta::Pipe, Some(self.as_raw_fd()))
    |                                            ^^^^^^^^^ method not found in `&ChildStderr`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `sys::unix::ext::io::AsRawFd` defines an item `as_raw_fd`, perhaps you need to implement it
    |
    |
22  | pub trait AsRawFd {

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0432, E0599.
Some errors have detailed explanations: E0432, E0599.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `std`

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:01:47
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1645178071) [(plain)](https://github.com/rust-lang/rust/commit/6a1cc44fd3f0d9466ef1e653904ff9fe4cbdfb04/checks/1645178071/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking hashbrown v0.9.0
    Checking object v0.22.0
    Checking miniz_oxide v0.4.0
    Checking addr2line v0.14.0
error: expected one of `!`, `(`, `+`, `::`, `<`, `where`, or `{`, found `;`
   --> library/std/src/sys/unix/ext/process.rs:212:39
    |
212 | impl private::Sealed for ExitStatusExt;
    |                                       ^ expected one of 7 possible tokens

error[E0432]: unresolved imports `super::process::CommandExt`, `super::process::ExitStatusExt`
   |
   |
74 |     pub use super::process::{CommandExt, ExitStatusExt};
   |                              ^^^^^^^^^^  ^^^^^^^^^^^^^ no `ExitStatusExt` in `sys::unix::ext::process`
   |                              |
   |                              no `CommandExt` in `sys::unix::ext::process`

error[E0599]: no method named `as_raw_fd` found for reference `&ChildStdin` in the current scope
   --> library/std/src/sys/unix/kernel_copy.rs:364:44
    |
364 |         CopyParams(FdMeta::Pipe, Some(self.as_raw_fd()))
    |                                            ^^^^^^^^^ method not found in `&ChildStdin`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `sys::unix::ext::io::AsRawFd` defines an item `as_raw_fd`, perhaps you need to implement it
    |
    |
22  | pub trait AsRawFd {


error[E0599]: no method named `as_raw_fd` found for reference `&ChildStdout` in the current scope
   --> library/std/src/sys/unix/kernel_copy.rs:370:44
    |
370 |         CopyParams(FdMeta::Pipe, Some(self.as_raw_fd()))
    |                                            ^^^^^^^^^ method not found in `&ChildStdout`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `sys::unix::ext::io::AsRawFd` defines an item `as_raw_fd`, perhaps you need to implement it
    |
    |
22  | pub trait AsRawFd {


error[E0599]: no method named `as_raw_fd` found for reference `&ChildStderr` in the current scope
   --> library/std/src/sys/unix/kernel_copy.rs:376:44
    |
376 |         CopyParams(FdMeta::Pipe, Some(self.as_raw_fd()))
    |                                            ^^^^^^^^^ method not found in `&ChildStderr`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `sys::unix::ext::io::AsRawFd` defines an item `as_raw_fd`, perhaps you need to implement it
    |
    |
22  | pub trait AsRawFd {

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0432, E0599.
Some errors have detailed explanations: E0432, E0599.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `std`

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:01:39
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1645238764) [(plain)](https://github.com/rust-lang/rust/commit/2db5aa2b3ab572284ad9c47650bceb6fbdf8fa46/checks/1645238764/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking addr2line v0.14.0
error: trait objects without an explicit `dyn` are deprecated
   --> library/std/src/sys/unix/ext/process.rs:212:26
    |
212 | impl private::Sealed for ExitStatusExt { }
    |                          ^^^^^^^^^^^^^ help: use `dyn`: `dyn ExitStatusExt`
    |
    = note: `-D bare-trait-objects` implied by `-D warnings`

error[E0038]: the trait `ExitStatusExt` cannot be made into an object
   --> library/std/src/sys/unix/ext/process.rs:212:26
    |
212 | impl private::Sealed for ExitStatusExt { }
    |                          ^^^^^^^^^^^^^ `ExitStatusExt` cannot be made into an object
    |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
   --> library/std/src/sys/unix/ext/process.rs:181:8
    |
177 | pub trait ExitStatusExt : private::Sealed {
    |           ------------- this trait cannot be made into an object...
...
181 |     fn from_raw(raw: i32) -> Self;
    |        ^^^^^^^^ ...because associated function `from_raw` has no `self` parameter
help: consider turning `from_raw` into a method by giving it a `&self` argument
    |
181 |     fn from_raw(&self, raw: i32) -> Self;
    |                 ^^^^^^
help: alternatively, consider constraining `from_raw` so it does not apply to trait objects
    |
181 |     fn from_raw(raw: i32) -> Self where Self: Sized;


error[E0277]: the trait bound `process::ExitStatus: Sealed` is not satisfied
   --> library/std/src/sys/unix/ext/process.rs:215:6
    |
177 | pub trait ExitStatusExt : private::Sealed {
    |                           --------------- required by this bound in `ExitStatusExt`
...
215 | impl ExitStatusExt for process::ExitStatus {
    |      ^^^^^^^^^^^^^ the trait `Sealed` is not implemented for `process::ExitStatus`
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0038, E0277.
For more information about an error, try `rustc --explain E0038`.
For more information about an error, try `rustc --explain E0038`.
error: could not compile `std`

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:01:29
```

</details>

ijackson commented:
@modify labels -S-waiting-on-review +S-waiting-on-author

ijackson commented:
@rustbot modify labels -S-waiting-on-review +S-waiting-on-author

ijackson commented:
Hi again.  Sorry for the flail on this branch while I tried to get the CI to tell my my syntax errors :-).  Worked better when I did it locally.

I think this is all sorted now, and I have squashed it a bit and rebased onto current master.

NB: I made the new trait `Sealed` insta-stable.  Please let me know if that's not right.

Thanks.

@rustbot modify labels -S-waiting-on-author +S-waiting-on-review

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 26739e45da5f24520e81a219e280bd6bcf723cfe has been approved by `dtolnay`

<!-- @bors r=dtolnay 26739e45da5f24520e81a219e280bd6bcf723cfe -->
<!-- homu: {"type":"Approved","sha":"26739e45da5f24520e81a219e280bd6bcf723cfe","approver":"dtolnay"} -->

bors commented:
:hourglass: Testing commit 26739e45da5f24520e81a219e280bd6bcf723cfe with merge 6c72503d98fd4c59d861e8d0e90d0f7f441ac3b9...
<!-- homu: {"type":"BuildStarted","head_sha":"26739e45da5f24520e81a219e280bd6bcf723cfe","merge_sha":"6c72503d98fd4c59d861e8d0e90d0f7f441ac3b9"} -->

rust-log-analyzer commented:
The job **`dist-various-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/1678786897) [(plain)](https://github.com/rust-lang-ci/rust/commit/6c72503d98fd4c59d861e8d0e90d0f7f441ac3b9/checks/1678786897/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] addr2line test:false 0.465
[RUSTC-TIMING] core test:false 32.197
[RUSTC-TIMING] gimli test:false 5.017
[RUSTC-TIMING] object test:false 9.358
error[E0599]: no method named `core_dumped` found for reference `&process_inner::ExitStatus` in the current scope
   --> library/std/src/sys/unix/ext/process.rs:227:25
    |
227 |         self.as_inner().core_dumped()
    |                         ^^^^^^^^^^^ method not found in `&process_inner::ExitStatus`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `ExitStatusExt` defines an item `core_dumped`, perhaps you need to implement it
   --> library/std/src/sys/unix/ext/process.rs:178:1
    |
178 | pub trait ExitStatusExt: private::Sealed {


error[E0599]: no method named `stopped_signal` found for reference `&process_inner::ExitStatus` in the current scope
   --> library/std/src/sys/unix/ext/process.rs:231:25
    |
231 |         self.as_inner().stopped_signal()
    |                         ^^^^^^^^^^^^^^ method not found in `&process_inner::ExitStatus`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `ExitStatusExt` defines an item `stopped_signal`, perhaps you need to implement it
   --> library/std/src/sys/unix/ext/process.rs:178:1
    |
178 | pub trait ExitStatusExt: private::Sealed {


error[E0599]: no method named `continued` found for reference `&process_inner::ExitStatus` in the current scope
   --> library/std/src/sys/unix/ext/process.rs:235:25
    |
235 |         self.as_inner().continued()
    |                         ^^^^^^^^^ method not found in `&process_inner::ExitStatus`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `ExitStatusExt` defines an item `continued`, perhaps you need to implement it
   --> library/std/src/sys/unix/ext/process.rs:178:1
    |
178 | pub trait ExitStatusExt: private::Sealed {


error[E0599]: no method named `into_raw` found for reference `&process_inner::ExitStatus` in the current scope
   --> library/std/src/sys/unix/ext/process.rs:239:25
    |
239 |         self.as_inner().into_raw().into()
    |                         ^^^^^^^^ method not found in `&process_inner::ExitStatus`
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `ExitStatusExt` defines an item `into_raw`, perhaps you need to implement it
   --> library/std/src/sys/unix/ext/process.rs:178:1
    |
178 | pub trait ExitStatusExt: private::Sealed {

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0599`.
For more information about this error, try `rustc --explain E0599`.
[RUSTC-TIMING] std test:false 2.217
error: could not compile `std`

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-fuchsia" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap dist --host= --target x86_64-fuchsia,aarch64-fuchsia,wasm32-unknown-unknown,wasm32-wasi,sparcv9-sun-solaris,x86_64-sun-solaris,x86_64-unknown-linux-gnux32,x86_64-fortanix-unknown-sgx,nvptx64-nvidia-cuda,armv7-unknown-linux-gnueabi,armv7-unknown-linux-musleabi,i686-unknown-freebsd
Build completed unsuccessfully in 0:14:51
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1678840816)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1678840816","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1694544204) [(plain)](https://github.com/rust-lang/rust/commit/7776840aa920e21ec47c8da641fad5f158635459/checks/1694544204/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/sys/unix/process/process_fuchsia.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Diff in /checkout/library/std/src/sys/unix/process/process_fuchsia.rs at line 254:
     // other things from std::os::unix) properly.  This veneer is always going to be a bdoge.  So
     // while I don't know if these implementations are actually correct, I think they will do for
     // now at least.
-    pub fn core_dumped(&self) -> bool { false }
-    pub fn stopped_signal(&self) -> Option<i32> { None }
-    pub fn continued(&self) -> bool { false }
+    pub fn core_dumped(&self) -> bool {
+    }
+    }
+    pub fn stopped_signal(&self) -> Option<i32> {
+        None
+    }
+    pub fn continued(&self) -> bool {
+    }
 
 
     pub fn into_raw(&self) -> c_int {
         // We don't know what someone who calls into_raw() will do with this value, but it should
Build completed unsuccessfully in 0:00:18
```

</details>

ijackson commented:
I have fixed the Fuchsia build (sadly the area seems a bit of a mess) and rebased this onto current master.

I wonder if this will work:
@bors try

bors commented:
@ijackson: :key: Insufficient privileges: not in try users

ijackson commented:
> @ijackson: key Insufficient privileges: not in try users

Oh well.

@rustbot modify labels -S-waiting-on-author +S-waiting-on-review

ijackson commented:
@rustbot modify labels -S-waiting-on-author +S-waiting-on-review

ijackson commented:
@dtolnay would you mind re-reviewing?  You might want to look at my "fix fuchsia" commit in particular since the situation is suboptimal to start with and I had to make some ... decisions.

As regards the build failure, I did a local "./x.py check" targeting fuchsia which found the original build failure as detected by bors and have checked that this commit fixes it.

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit a8d01619608715e6abc4c6d3c6f347f393262725 has been approved by `dtolnay`

<!-- @bors r=dtolnay a8d01619608715e6abc4c6d3c6f347f393262725 -->
<!-- homu: {"type":"Approved","sha":"a8d01619608715e6abc4c6d3c6f347f393262725","approver":"dtolnay"} -->

dtolnay commented:
I added a note in the tracking issue to run this by a Fuchsia dev prior to stabilization.
